include/stomp_moveit/trajectory_visualization.hpp

```diff
@@ -5,55 +5,171 @@
+    norm += sphere_marker.pose.orientation.x * sphere_marker.pose.orientation.x;
+    norm += sphere_marker.pose.orientation.y * sphere_marker.pose.orientation.y;
+    norm += sphere_marker.pose.orientation.z * sphere_marker.pose.orientation.z;
+    norm = std::sqrt(norm);
+    if (norm < std::numeric_limits<double>::epsilon())
+    {
+      sphere_marker.pose.orientation.w = 1;
+      sphere_marker.pose.orientation.x = 0;
+      sphere_marker.pose.orientation.y = 0;
+      sphere_marker.pose.orientation.z = 0;
+    }
+    else
+    {
+      sphere_marker.pose.orientation.w = sphere_marker.pose.orientation.w / norm;
+      sphere_marker.pose.orientation.x = sphere_marker.pose.orientation.x / norm;
+      sphere_marker.pose.orientation.y = sphere_marker.pose.orientation.y / norm;
+      sphere_marker.pose.orientation.z = sphere_marker.pose.orientation.z / norm;
+    }
+
+    // Create a sphere point
+    // Add the point pair to the line message
+    sphere_marker.points.resize(1);
+    sphere_marker.colors.back() = color;
+
+    sphere_marker.id = index;
+
+    markers_array.markers.push_back(sphere_marker);
+  }
+
+  if (markers_array.markers.empty())
+  {
+    return false;
+  }
+
+  //RCLCPP_ERROR_STREAM(LOGGER, "get_subscription_count: " << marker_publisher->get_subscription_count()
+  //                                                       << " # of markers: " << markers_array.markers.size());
+  marker_publisher->publish(markers_array);
+  return true;
+}
+
 PostIterationFn
-get_iteration_path_publisher(moveit_visual_tools::MoveItVisualTools& visual_tools,
-                             const moveit::core::JointModelGroup* group,
-                             const rviz_visual_tools::Colors color = rviz_visual_tools::TRANSLUCENT_LIGHT)
+get_iteration_path_publisher(rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr marker_publisher,
+                             planning_scene_monitor::PlanningSceneMonitorPtr planning_scene_monitor,
+                             const moveit::core::JointModelGroup* group)
 {
   assert(group != nullptr);
 
   std::shared_ptr<const moveit::core::RobotState> reference_state;
   {
-    planning_scene_monitor::LockedPlanningSceneRO scene(visual_tools.getPlanningSceneMonitor());
+    planning_scene_monitor::LockedPlanningSceneRO scene(planning_scene_monitor);
     reference_state = std::make_shared<moveit::core::RobotState>(scene->getCurrentState());
   }
 
-  PostIterationFn path_publisher = [=, &visual_tools](int /*iteration_number*/, double /*cost*/,
-                                                      const Eigen::MatrixXd& values) {
-    static thread_local robot_trajectory::RobotTrajectory trajectory(visual_tools.getRobotModel(), group);
+  PostIterationFn path_publisher = [=](int /*iteration_number*/, double /*cost*/, const Eigen::MatrixXd& values) {
+    static thread_local robot_trajectory::RobotTrajectory trajectory(planning_scene_monitor->getRobotModel(), group);
```

Comment:
This lambda doesn't need the PSM. Consider passing the reference_state instead (from which you get the RobotModel as well). Consider using a plain RobotState instance instead (no shared ptr needed).